src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,225 @@
+#include <seastar/core/gate.hh>
+#include <seastar/core/internal/pollable_fd.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/later.hh>
+
+#include <boost/lockfree/spsc_queue.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+#include <mutex>
+
+namespace v8_engine {
+
+namespace internal {
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() noexcept = 0;
+    virtual void cancel() noexcept = 0;
+    virtual void on_timeout() noexcept = 0;
+    virtual void set_exception(const std::exception_ptr& exception) {
+        _exception = exception;
+    };
+    virtual std::chrono::milliseconds get_timeout() const noexcept = 0;
+
+    virtual void done() = 0;
+
+    std::exception_ptr _exception;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    task(Func&& f, std::chrono::milliseconds timeout)
+      : work_item()
+      , _func(std::forward<Func>(f))
+      , _timeout(timeout) {}
+
+    // Save first exception for task
+    template<typename FunForExecute>
+    void func_wrapper(FunForExecute&& func) {
+        if (_exception) {
+            return;
+        }
+
+        try {
+            func();
+        } catch (...) {
+            set_exception(std::current_exception());
+        }
+    }
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() noexcept override {
+        func_wrapper([this] { _func(); });
+    }
+
+    // Run this function for cancel task execution
+    void cancel() noexcept override {
+        func_wrapper([this] { _func.cancel(); });
+    }
+
+    // Run this function after wathdog was alarmed and execution was canceled.
+    // For example: prepare inside task fields for next execution
+    void on_timeout() noexcept override {
+        func_wrapper([this] { _func.on_timeout(); });
+    }
+
+    std::chrono::milliseconds get_timeout() const noexcept override {
+        return _timeout;
+    }
+
+    // Signal from std::thread to seastar about the end of task processing
+    void done() override { _on_done.write_side().signal(1); }
+
+    // Get future from seastar thread for waiting when executor thread will
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().discard_result().then([this] {
+            if (_exception) {
+                return ss::make_exception_future(_exception);
+            } else {
+                return ss::now();
+            }
+        });
+    }
+
+    Func _func;
+    seastar::readable_eventfd _on_done;
+
+    std::chrono::milliseconds _timeout;
+};
+
+// This class implement queue for submit task from seastar to std::thread. Only
+// one core with seastar can submit task to queue. Only one std::thread can
+// consume task from queue.
+class spsc_queue {
```

Comment:
BTW, why it's called spsc? You have multiple producers and one consumer. The doc comment above says that `Only one core with seastar can submit task to queue` but as I understand, any shard can submit but only one at a time.

Reply:
I have only one producer core. We need to use invoke_on for executor and create it in one core